### ISSUE TITLE
Creating a Caffeine LoadingCache using a Guava CacheLoader

### ISSUE THREAD

**Issue by Zuplyx** [ben-manes/caffeine#766]
Hi,

we are currently in the process of migrating from Guava cache to Caffeine. In our codebase we have several classes implementing Guava's CacheLoader. Ideally we'd like to leave these classes unchanged.
Is there a way to create a Caffeine LoadingCache using a Guava CacheLoader? 
Maybe an analogue to CaffeinatedGuava's [build](https://www.javadoc.io/static/com.github.ben-manes.caffeine/guava/3.1.1/com.github.benmanes.caffeine.guava/com/github/benmanes/caffeine/guava/CaffeinatedGuava.html#build(com.github.benmanes.caffeine.cache.Caffeine,com.google.common.cache.CacheLoader)) method, but which directly returns the created Caffeine cache instead of wrapping it in a Guava cache?

**Comment by ben-manes** [comment]
I suppose it depends on which features of Guava's CacheLoader you are implementing. I think you could write an adapter without too much trouble.

### Single loading
This is a 1-to-1 so you can use a method reference
```java
LoadingCache<K, V> cache = Caffeine.newBuilder().build(guavaCacheLoader::load);
```

### Bulk loading
This is slightly harder since Caffeine detects if `loadAll` is implemented at construction, so you should be mindful to use this only when appropriate. Otherwise a bulk `getAll` will call Guava's default method to throw an exception, which is not handled at runtime due to the prior detection. This differs from Guava which swallows the exception at runtime to fallback to singular loads.

```java
LoadingCache<K, V> cache = Caffeine.newBuilder().build(new CacheLoader<K, V>() {
  public K load(K key) {
    return guavaCacheLoader.load(key);
  }
  public Map<K, V> loadAll(Set<? extends K> keys) {
    return guavaCacheLoader.loadAll(keys);
  }
});
```

If you don't mind using the bulk `loadAll` for singular lookups, then you could simplify this to a method reference.
```java
LoadingCache<K, V> cache = Caffeine.newBuilder().build(CacheLoader.bulk(guavaCacheLoader::loadAll));
```

### Reloading
If you customize the `reload(key)` method to use the old value, then the futures need to be converted. The [future-converter](https://github.com/lukas-krecan/future-converter) library makes this easier.

```java
import net.javacrumbs.futureconverter.java8guava.FutureConverter;

LoadingCache<K, V> cache = Caffeine.newBuilder().build(new CacheLoader<K, V>() {
  public K load(K key) {
    return guavaCacheLoader.load(key);
  }
  public CompletableFuture<? extends V> asyncReload(K key, V oldValue, Executor executor) {
    return FutureConverter.toCompletableFuture(guavaCacheLoader.reload(key, oldValue));
  }
});
```

**Comment by ben-manes** [comment]
I've exposed our existing adapters in the Guava package. I'm not sure when the next release will be, so you might review that code to embed something similar locally. It's a little more complex since it handles single, bulk, async loading, and some internal adaptation details.

```java
CacheLoader<K, V> caffeineLoader = CaffeinatedGuava.caffeinate(guavaLoader);
LoadingCache<K, V> caffeineCache = Caffeine.newBuilder().build(caffeineLoader);
```

**Comment by Zuplyx** [comment]
Thanks Ben. That's exactly what we needed.

**Comment by patric-r** [comment]
We're facing exactly the same situation as @Zuplyx.
We're in the transitioning phase from guava cache to your awesome Caffeine but we also have many Guava CacheLoader implementations.
However, this issue (including loadAll/reload feature which we heavily use as well ) prevents us from switching to caffeine. 

@ben-manes This issue has been fixed more than two months ago. Any chance to get a release out soon?
I'm sure this will be an awesome improvement and helps many projects migrating to Caffeine from Guava.

Thanks and keep up the great work!

**Comment by ben-manes** [comment]
Released in 3.1.2
